core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -4044,6 +4049,52 @@ private OutputColumn createOutputColumn(Field field)
         {
             return new OutputColumn(new Column(field.getName().orElseThrow(), field.getType().toString()), analysis.getSourceColumns(field));
         }
+
+        /**
+         * Helper function that analyzes any versioning and returns the appropriate table handle.
+         * If no for clause exists, this is just a wrapper around getRedirectionAwareTableHandle in MetadataManager.
+         */
+        private RedirectionAwareTableHandle getTableHandle(Table table, QualifiedObjectName name, Optional<Scope> scope)
+        {
+            if (table.getQueryPeriod().isPresent()) {
+                Optional<TableVersion> tableVersion = extractTableVersion(table, scope);
+                Optional<TableVersionHandle> tableVersionHandle = metadata.getVersionHandle(session, name, tableVersion.get());
+                analysis.setTableVersion(table, tableVersionHandle.get());
+
+                return metadata.getRedirectionAwareTableHandle(session, name, tableVersionHandle);
+            }
+            return metadata.getRedirectionAwareTableHandle(session, name);
+        }
+
+        /**
+         * Analyzes a for clause and extracts an evaluated version value
+         */
+        private Optional<TableVersion> extractTableVersion(Table table, Optional<Scope> scope)
+        {
+            if (!table.getQueryPeriod().isPresent()) {
+                return Optional.empty();
+            }
+            Optional<TableVersion> tableVersion = Optional.empty();
+            Expression version = table.getQueryPeriod().get().getRangeExpression();
+
+            ExpressionAnalysis expressionAnalysis = analyzeExpression(version, scope.get());
+            analysis.recordSubqueries(table, expressionAnalysis);
+
+            // Once the range value is analyzed, we can evaluate it
+            Type versionType = expressionAnalysis.getType(version);
+            TravelType travelType = getTravelTypeFromRangeType(table.getQueryPeriod().get().getRangeType());
+            Object evaluatedVersion = evaluateConstantExpression(version, versionType, metadata, session, accessControl, ImmutableMap.of());
```

Comment:
If travel type is SYSTEM_TIME, the type of the expression should be TIMESTAMP WITH TIME ZONE.

We should discuss what it should be in the case of SYSTEM_VERSION.

The type should be checked at this point during analysis and a proper error should be thrown if the type is not valid for the given travel type (i.e., a TrinoException with TYPE_MISMATCH).

Reply:
Discussed this a little further up when talking about version type, but I ended up creating SPIs for the versions so the connectors can determine what data types are valid. Given the various data types that different data lake solutions support for time travel, I thought this was a good way to allow connectors to enforce typing. Let me know if you want to talk about it more.